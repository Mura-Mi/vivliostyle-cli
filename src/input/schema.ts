import * as v from 'valibot';

export type StructuredDocument = {
  title: string;
  href: string;
  sections?: StructuredDocumentSection[];
  children: StructuredDocument[];
};

export type StructuredDocumentSection = {
  headingHtml: string;
  headingText: string;
  level: number;
  href?: string;
  id?: string;
  children: StructuredDocumentSection[];
};

export const ValidString = v.pipe(
  v.string(),
  v.trim(),
  v.minLength(1, 'At least one character is required'),
);

export const ThemeObject = v.intersect([
  v.required(
    v.object({
      /**
       * Specifier name of importing theme package or a path of CSS file.
       * - A npm-style package argument is allowed (ex: @vivliostyle/theme-academic@1 ./local-pkg)
       * - A URL or a local path of CSS is allowed (ex: ./style.css, https://example.com/style.css)
       */
      specifier: ValidString,
    }),
    'Missing required field: specifier',
  ),
  v.partial(
    v.object({
      /**
       * Importing CSS path(s) of the package. Specify this if you want to import other than the default file.
       */
      import: v.union([v.array(ValidString), ValidString]),
    }),
  ),
]);
export type ThemeObject = v.InferInput<typeof ThemeObject>;

export const ThemeSpecifier = v.union([
  v.array(v.union([ThemeObject, ValidString])),
  ThemeObject,
  ValidString,
]);
export type ThemeSpecifier = v.InferInput<typeof ThemeSpecifier>;

export const ArticleEntryObject = v.object({
  path: ValidString,
  output: v.optional(ValidString),
  title: v.optional(ValidString),
  theme: v.optional(ThemeSpecifier),
  encodingFormat: v.optional(ValidString),
  rel: v.optional(v.union([v.array(ValidString), ValidString])),
});
export type ArticleEntryObject = v.InferInput<typeof ArticleEntryObject>;

const PageBreak = v.union([
  v.literal('left'),
  v.literal('right'),
  v.literal('recto'),
  v.literal('verso'),
]);

export const ContentsEntryObject = v.object({
  rel: v.literal('contents'),
  path: v.optional(ValidString),
  output: v.optional(ValidString),
  title: v.optional(ValidString),
  theme: v.optional(ThemeSpecifier),
  /**
   * Specify the page break position before this document. It is useful when you want to specify which side a first page of the document should be placed on a two-page spread.
   */
  pageBreakBefore: v.optional(PageBreak),
  /**
   * Reset the starting page number of this document by the specified integer. It is useful when you want to control page numbers when including a
   *  page.
   */
  pageCounterReset: v.optional(v.pipe(v.number(), v.safeInteger())),
});
export type ContentsEntryObject = v.InferInput<typeof ContentsEntryObject>;

export const CoverEntryObject = v.object({
  rel: v.literal('cover'),
  path: v.optional(ValidString),
  output: v.optional(ValidString),
  title: v.optional(ValidString),
  theme: v.optional(ThemeSpecifier),
  imageSrc: v.optional(ValidString),
  imageAlt: v.optional(v.string()), // Allow empty string
  /**
   * Specify the page break position before this document. It is useful when you want to specify which side a first page of the document should be placed on a two-page spread.
   */
  pageBreakBefore: v.optional(PageBreak),
});
export type CoverEntryObject = v.InferInput<typeof CoverEntryObject>;

export const EntryObject = v.union([
  ContentsEntryObject,
  CoverEntryObject,
  v.required(ArticleEntryObject, ['path'], 'Missing required field: path'),
]);
export type EntryObject = v.InferInput<typeof EntryObject>;

export const OutputObject = v.intersect([
  v.required(
    v.object({
      /**
       * Specify output file name or directory [<title>.pdf].
       */
      path: ValidString,
    }),
    'Missing required field: path',
  ),
  v.partial(
    v.object({
      /**
       * Specify output format.
       */
      format: ValidString,
      /**
       * if docker is set, Vivliostyle try to render PDF on Docker container [local].
       */
      renderMode: v.union([v.literal('local'), v.literal('docker')]),
      /**
       * Apply the process to generate PDF for printing.
       */
      preflight: v.union([
        v.literal('press-ready'),
        v.literal('press-ready-local'),
      ]),
      /**
       * Options for preflight process (ex: gray-scale, enforce-outline). Please refer the document of press-ready for further information. https://github.com/vibranthq/press-ready
       */
      preflightOption: v.array(ValidString),
    }),
  ),
]);
export type OutputObject = v.InferInput<typeof OutputObject>;

// Use v.looseObject to allow unknown keys in future VFM versions
export const VfmReplaceRule = v.looseObject({
  test: v.instance(RegExp),
  match: v.function() as v.GenericSchema<
    (result: RegExpMatchArray, h: any) => Object | string
  >,
});
export type VfmReplaceRule = v.InferInput<typeof VfmReplaceRule>;

export const BrowserType = v.union([
  v.literal('chromium'),
  v.literal('firefox'),
  v.literal('webkit'),
]);
export type BrowserType = v.InferInput<typeof BrowserType>;

export const VivliostyleConfigEntry = v.intersect([
  v.required(
    v.object({
      /**
       * Entry file(s) of document.
       */
      entry: v.union([
        v.pipe(
          v.array(v.union([ValidString, EntryObject])),
          v.minLength(1, 'At least one entry is required'),
        ),
        v.required(
          ArticleEntryObject,
          ['path'],
          'Missing required field: path',
        ),
        ValidString,
      ]),
    }),
    'Missing required field: entry',
  ),
  v.partial(
    v.object({
      /**
       * Title
       */
      title: ValidString,
      /**
       * Author
       */
      author: ValidString,
      /**
       * Theme package path(s) or URL(s) of css file.
       */
      theme: ThemeSpecifier,
      /**
       * Directory of referencing entry file(s).
       */
      entryContext: ValidString,
      /**
       * Options about outputs.
       */
      output: v.union([
        v.array(v.union([OutputObject, ValidString])),
        OutputObject,
        ValidString,
      ]),
      /**
       * Specify the directory where the intermediate files (manuscript HTMLs, publication.json, etc.) are saved.
       * If not specified, theses files are saved in the same directory as the input file.
       */
      workspaceDir: ValidString,
      /**
       * @deprecated Use 'copyAsset.includes' instead
       */
      includeAssets: v.union([v.array(ValidString), ValidString]),
      /**
       * Options about asset files to be copied when exporting output.
       */
      copyAsset: v.partial(
        v.object({
          /**
           * Specify directories and files you want to include as asset files. This option supports wildcard characters to make glob patterns.
           */
          includes: v.array(ValidString),
          /**
           * Specify directories and files you want to exclude from the asset file. This option supports wildcard characters to make glob patterns.
           */
          excludes: v.array(ValidString),
          /**
           * Specify extensions of the file you want to include as an asset file. (default: [png, jpg, jpeg, svg, gif, webp, apng, ttf, otf, woff, woff2])
           */
          includeFileExtensions: v.array(ValidString),
          /**
           * Specify extensions of the file you want to exclude as an asset file.
           */
          excludeFileExtensions: v.array(ValidString),
        }),
      ),
      /**
       * Output pdf size [Letter]. preset: A5, A4, A3, B5, B4, JIS-B5, JIS-B4, letter, legal, ledger. custom(comma separated): 182mm,257mm or 8.5in,11in.
       */
      size: ValidString,
      /**
       * Make generated PDF compatible with press ready PDF/X-1a [false]. This option is equivalent with "preflight": "press-ready"
       */
      pressReady: v.boolean(),
      /**
       * Language
       */
      language: ValidString,
      readingProgression: v.union([v.literal('ltr'), v.literal('rtl')]),
      /**
       * Options about Table of Contents (ToC) documents.
       */
      toc: v.union([
        v.partial(
          v.object({
            /**
             * Specify the title of the generated ToC document.
             */
            title: ValidString,
            /**
             * Specify the location where the generated ToC document will be saved. (default: index.html)
             */
            htmlPath: ValidString,
            /**
             * Specify the depth of the section to be included in the ToC document. (default: 0)
             */
            sectionDepth: v.pipe(
              v.number(),
              v.integer(),
              v.minValue(0),
              v.maxValue(6),
            ),
            /**
             * Specify the transform function for the document list.
             */
            transformDocumentList: v.function() as v.GenericSchema<
              (
                nodeList: StructuredDocument[],
              ) => (propsList: { children: any }[]) => any
            >,
            /**
             * Specify the transform function for the section list.
             */
            transformSectionList: v.function() as v.GenericSchema<
              (
                nodeList: StructuredDocumentSection[],
              ) => (propsList: { children: any }[]) => any
            >,
          }),
        ),
        v.boolean(),
        ValidString,
      ]),
      /**
       * @deprecated Use 'toc.title' instead
       */
      tocTitle: ValidString,
      /**
       * Options about cover images and cover page documents
       */
      cover: v.union([
        v.intersect([
          v.required(
            v.object({
              /**
               * Specify the cover image to be used for the cover page.
               */
              src: ValidString,
            }),
            'Missing required field: src',
          ),
          v.partial(
            v.object({
              /**
               * Specify alternative text for the cover image.
               */
              name: v.string(), // Allow empty string
              /**
               * Specify the location where the generated cover document will be saved. (default: cover.html) If falsy value is set, the cover document won't be generated.
               */
              htmlPath: v.union([ValidString, v.boolean()]),
            }),
          ),
        ]),
        ValidString,
      ]),
      /**
       * Timeout limit for waiting Vivliostyle process [120000]
       */
      timeout: v.number(),
      /**
       * Option for convert Markdown to a stringify (HTML).
       */
      vfm: v.partial(
        // Use v.looseObject to allow unknown keys in future VFM versions
        v.looseObject({
          /**
           * Custom stylesheet path/URL.
           */
          style: v.union([v.array(ValidString), ValidString]),
          /**
           * Output markdown fragments.
           */
          partial: v.boolean(),
          /**
           * Document title (ignored in partial mode).
           */
          title: ValidString,
          /**
           * Document language (ignored in partial mode).
           */
          language: ValidString,
          /**
           * Replacement handler for HTML string.
           */
          replace: v.array(VfmReplaceRule),
          /**
           * Add `<br>` at the position of hard line breaks, without needing spaces.
           */
          hardLineBreaks: v.boolean(),
          /**
           * Disable automatic HTML format.
           */
          disableFormatHtml: v.boolean(),
          /**
           * Enable math syntax.
           */
          math: v.boolean(),
        }),
      ),
      /**
       * Specify a docker image to render.
       */
      image: ValidString,
      /**
       * Launch an HTTP server hosting contents instead of file protocol. It is useful that requires CORS such as external web fonts.
       */
      http: v.boolean(),
      /**
       * Specify a URL of displaying viewer instead of vivliostyle-cli's one. It is useful that using own viewer that has staging features. (ex: https://vivliostyle.vercel.app/)
       */
      viewer: ValidString,
      /**
       * specify viewer parameters. (ex: "allowScripts=false&pixelRatio=16")
       */
      viewerParam: ValidString,
      /**
       * EXPERIMENTAL SUPPORT: Specify a browser type to launch Vivliostyle viewer. Currently, Firefox and Webkit support preview command only!
       */
      browser: BrowserType,
    }),
  ),
]);
export type VivliostyleConfigEntry = v.InferInput<
  typeof VivliostyleConfigEntry
>;

export const VivliostyleConfigSchema = v.union([
  v.pipe(
    v.array(VivliostyleConfigEntry),
    v.minLength(1, 'At least one config entry is required'),
  ),
  VivliostyleConfigEntry,
]);
export type VivliostyleConfigSchema = v.InferInput<
  typeof VivliostyleConfigSchema
>;
