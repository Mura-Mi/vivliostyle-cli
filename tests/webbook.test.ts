import { fs as memfs, vol } from 'memfs';
import { format } from 'prettier';
import { afterEach, expect, it, vi } from 'vitest';
import { build } from '../src/index.js';
import { VivliostyleConfigSchema } from '../src/schema/vivliostyleConfig.schema.js';
import { toTree } from './commandUtil.js';

vi.mock('node:fs', () => ({ ...memfs, default: memfs }));

vi.mock('@vivliostyle/jsdom', () =>
  import('./commandUtil.js').then(({ getMockedJSDOM }) => getMockedJSDOM()),
);

vi.mock('../src/processor/theme.ts', async (importOriginal) => ({
  ...(await importOriginal<typeof import('../src/processor/theme.js')>()),
  checkThemeInstallationNecessity: () => Promise.resolve(false),
  installThemeDependencies: () => Promise.resolve(),
}));

afterEach(() => vol.reset());

it('generate webpub from single markdown file', async () => {
  vol.fromJSON({
    '/work/input/foo.md': '# Hi',
  });
  await build({
    input: '/work/input/foo.md',
    targets: [{ path: '/work/output', format: 'webpub' }],
  });

  expect(toTree(vol)).toMatchSnapshot();
  const file = vol.toJSON();
  const manifest = JSON.parse(file['/work/output/publication.json'] as string);
  delete manifest.dateModified;
  expect(manifest).toMatchSnapshot();
});

it('generate webpub from vivliostyle.config.js', async () => {
  const config: VivliostyleConfigSchema = {
    entry: ['doc/one.md', 'doc/two.md'],
    output: ['/work/input/output', '/work/output'],
    toc: true,
    cover: 'cover.png',
    readingProgression: 'rtl',
  };
  vol.fromJSON({
    '/work/input/vivliostyle.config.json': JSON.stringify(config),
    '/work/input/doc/one.md': 'yuno',
    '/work/input/doc/two.md': 'yunocchi',
    '/work/input/cover.png': '',
  });
  await build({
    configPath: '/work/input/vivliostyle.config.json',
  });

  expect(toTree(vol)).toMatchSnapshot();
  const file = vol.toJSON();
  const manifest = JSON.parse(file['/work/output/publication.json'] as string);
  delete manifest.dateModified;
  expect(manifest).toMatchSnapshot();
  const toc = file['/work/output/index.html'];
  expect(await format(toc as string, { parser: 'html' })).toMatchSnapshot();

  const manifest2 = JSON.parse(
    file['/work/input/output/publication.json'] as string,
  );
  delete manifest2.dateModified;
  expect(manifest2).toEqual(manifest);
  const toc2 = file['/work/input/output/index.html'];
  expect(toc2).toEqual(toc);
});

it('generate webpub from a plain HTML', async () => {
  vol.fromJSON({
    '/work/input/webbook.html': /* html */ `
      <!DOCTYPE html>
      <html lang="ja">
      <head>
        <link rel="stylesheet" type="text/css" href="style.css">
        <title>Doc title</title>
      </head>
      <body>
      </body>
      </html>
    `,
    '/work/input/style.css': '',
  });
  await build({
    input: '/work/input/webbook.html',
    targets: [{ path: '/work/output', format: 'webpub' }],
  });

  expect(toTree(vol)).toMatchSnapshot();
  const file = vol.toJSON();
  const manifest = JSON.parse(file['/work/output/publication.json'] as string);
  delete manifest.dateModified;
  expect(manifest).toMatchSnapshot();
  const entry = file['/work/output/webbook.html'];
  expect(await format(entry as string, { parser: 'html' })).toMatchSnapshot();
});

it('generate webpub from a single-document publication', async () => {
  vol.fromJSON({
    '/work/input/webbook.html': /* html */ `
      <html lang="en">
      <head>
        <title>Document with toc</title>
        <link rel="publication" href="#wp_manifest">
        <script type="application/ld+json" id="wp_manifest">
          ${JSON.stringify({
            '@context': [
              'https://schema.org',
              'https://www.w3.org/ns/wp-context',
            ],
            conformsTo: 'yuno',
            resources: ['assets/figure.svg'],
            readingOrder: ['#foo', '../bar.html', 'subdir/index.html'],
          })}
        </script>
      </head>
      <body>

      </body>
      </html>
    `,
    '/work/input/subdir/index.html': /* html */ `
      <!DOCTYPE html>
      <html lang="ja">
      <head>
        <title>Doc title</title>
        <link rel="stylesheet" href="../assets/subdir.css">
      </head>
      <body>
      </body>
      </html>
    `,
    '/work/input/assets/figure.svg': '<svg></svg>',
    '/work/input/assets/subdir.css': '',
  });
  await build({
    input: '/work/input/webbook.html',
    targets: [{ path: '/work/output', format: 'webpub' }],
  });

  expect(toTree(vol)).toMatchSnapshot();
  const file = vol.toJSON();
  const entry = file['/work/output/webbook.html'];
  expect(await format(entry as string, { parser: 'html' })).toMatchSnapshot();
});

it('generate webpub from a remote HTML document', async () => {
  vol.fromJSON({
    '/work/input/index.html': /* html */ `
      <html lang="en">
      <head>
        <title>Document</title>
        <link rel="stylesheet" type="text/css" href="../assets/style.css">
      </head>
      <body>
      </body>
      </html>
    `,
    '/work/assets/style.css': '',
  });
  await build({
    input: 'https://example.com/work/input',
    targets: [{ path: '/work/output', format: 'webpub' }],
  });
  expect(toTree(vol)).toMatchSnapshot();
  const file = vol.toJSON();
  const manifest = JSON.parse(file['/work/output/publication.json'] as string);
  delete manifest.dateModified;
  expect(manifest).toMatchSnapshot();
  const entry = file['/work/output/work/input/index.html'];
  expect(await format(entry as string, { parser: 'html' })).toMatchSnapshot();
});

it('generate webpub with complex copyAsset settings', async () => {
  const config: VivliostyleConfigSchema = {
    copyAsset: {
      includes: [
        // The following line should also work, but under the mock environment of memfs, fast-glob does not work without '/**'.
        // 'node_modules/pkgB',
        'node_modules/pkgB/**',
      ],
      excludes: ['cover.png'],
      includeFileExtensions: ['jxl'],
      excludeFileExtensions: ['svg'],
    },
    entry: ['doc.md'],
    output: ['/work/output'],
    workspaceDir: '.vs',
  };
  vol.fromJSON({
    '/work/input/vivliostyle.config.json': JSON.stringify(config),
    '/work/input/package.json': '',
    '/work/input/doc.md': 'yuno',
    '/work/input/cover.png': '',
    '/work/input/assetA.jxl': '',
    '/work/input/assetB.svg': '',
    '/work/input/node_modules/pkgA/img.png': '',
    '/work/input/node_modules/pkgB/a.html': '',
    '/work/input/node_modules/pkgB/bar/b.html': '',
  });
  await build({
    configPath: '/work/input/vivliostyle.config.json',
  });

  expect(toTree(vol)).toMatchSnapshot();
});

it('copy webpub assets properly', async () => {
  const config: VivliostyleConfigSchema = {
    entry: ['doc.md'],
    output: ['/work/input/output1', '/work/input/output2'],
    theme: ['themeA', '@org/themeB'],
  };
  vol.fromJSON({
    '/work/input/vivliostyle.config.json': JSON.stringify(config),
    '/work/input/package.json': '',
    '/work/input/doc.md': 'yuno',
    '/work/input/node_modules/pkgA/a.html': '',
    '/work/input/node_modules/pkgA/a.css': '',
    '/work/input/themes/packages/themeA/package.json': '{"main": "theme.css"}',
    '/work/input/themes/packages/themeA/theme.css': '',
    '/work/input/themes/packages/themeA/example/a.css': '',
    '/work/input/themes/packages/@org/themeB/package.json':
      '{"main": "theme.css"}',
    '/work/input/themes/packages/@org/themeB/theme.css': '',
    '/work/input/themes/packages/@org/themeB/example/a.css': '',
  });
  await build({
    configPath: '/work/input/vivliostyle.config.json',
  });

  expect(toTree(vol)).toMatchSnapshot();
});
