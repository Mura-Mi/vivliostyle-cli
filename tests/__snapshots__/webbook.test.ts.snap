// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`copy webpub assets properly 1`] = `
"/
└─ work/
   └─ input/
      ├─ doc.html
      ├─ doc.md
      ├─ node_modules/
      │  └─ pkgA/
      │     ├─ a.css
      │     └─ a.html
      ├─ output1/
      │  ├─ doc.html
      │  ├─ publication.json
      │  └─ themes/
      │     └─ packages/
      │        ├─ @org/
      │        │  └─ themeB/
      │        │     └─ theme.css
      │        └─ themeA/
      │           └─ theme.css
      ├─ output2/
      │  ├─ doc.html
      │  ├─ publication.json
      │  └─ themes/
      │     └─ packages/
      │        ├─ @org/
      │        │  └─ themeB/
      │        │     └─ theme.css
      │        └─ themeA/
      │           └─ theme.css
      ├─ package.json
      ├─ publication.json
      ├─ themes/
      │  └─ packages/
      │     ├─ @org/
      │     │  └─ themeB/
      │     │     ├─ example/
      │     │     │  └─ a.css
      │     │     ├─ package.json
      │     │     └─ theme.css
      │     └─ themeA/
      │        ├─ example/
      │        │  └─ a.css
      │        ├─ package.json
      │        └─ theme.css
      └─ vivliostyle.config.json"
`;

exports[`generate webpub from a plain HTML 1`] = `
"/
└─ work/
   ├─ input/
   │  ├─ style.css
   │  └─ webbook.html
   └─ output/
      ├─ publication.json
      ├─ style.css
      └─ webbook.html"
`;

exports[`generate webpub from a plain HTML 2`] = `
Object {
  "@context": Array [
    "https://schema.org",
    "https://www.w3.org/ns/pub-context",
  ],
  "conformsTo": "https://github.com/vivliostyle/vivliostyle-cli",
  "inLanguage": "ja",
  "links": Array [],
  "name": "Doc title",
  "readingOrder": Array [
    Object {
      "url": "webbook.html",
    },
  ],
  "resources": Array [
    "style.css",
  ],
  "type": "Book",
}
`;

exports[`generate webpub from a plain HTML 3`] = `
"<!doctype html>
<html lang=\\"ja\\">
  <head>
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"style.css\\" />
    <title>Doc title</title>
    <link
      rel=\\"publication\\"
      type=\\"application/ld+json\\"
      href=\\"publication.json\\"
    />
  </head>
  <body></body>
</html>
"
`;

exports[`generate webpub from a remote HTML document 1`] = `
"/
└─ work/
   ├─ assets/
   │  └─ style.css
   ├─ input/
   │  └─ index.html
   └─ output/
      ├─ publication.json
      └─ work/
         ├─ assets/
         │  └─ style.css
         └─ input/
            └─ index.html"
`;

exports[`generate webpub from a remote HTML document 2`] = `
Object {
  "@context": Array [
    "https://schema.org",
    "https://www.w3.org/ns/pub-context",
  ],
  "conformsTo": "https://github.com/vivliostyle/vivliostyle-cli",
  "inLanguage": "en",
  "links": Array [],
  "name": "Document",
  "readingOrder": Array [
    Object {
      "url": "work/input/index.html",
    },
  ],
  "resources": Array [
    "work/assets/style.css",
  ],
  "type": "Book",
}
`;

exports[`generate webpub from a remote HTML document 3`] = `
"<html lang=\\"en\\">
  <head>
    <title>Document</title>
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"../assets/style.css\\" />
    <link
      rel=\\"publication\\"
      type=\\"application/ld+json\\"
      href=\\"../../publication.json\\"
    />
  </head>
  <body></body>
</html>
"
`;

exports[`generate webpub from a single-document publication 1`] = `
"/
└─ work/
   ├─ input/
   │  ├─ assets/
   │  │  ├─ figure.svg
   │  │  └─ subdir.css
   │  ├─ subdir/
   │  │  └─ index.html
   │  └─ webbook.html
   └─ output/
      ├─ assets/
      │  ├─ figure.svg
      │  └─ subdir.css
      ├─ subdir/
      │  └─ index.html
      └─ webbook.html"
`;

exports[`generate webpub from a single-document publication 2`] = `
"<html lang=\\"en\\">
  <head>
    <title>Document with toc</title>
    <link rel=\\"publication\\" href=\\"#wp_manifest\\" />
    <script type=\\"application/ld+json\\" id=\\"wp_manifest\\">
      {
        \\"@context\\": [\\"https://schema.org\\", \\"https://www.w3.org/ns/wp-context\\"],
        \\"conformsTo\\": \\"yuno\\",
        \\"resources\\": [\\"assets/figure.svg\\"],
        \\"readingOrder\\": [\\"#foo\\", \\"../bar.html\\", \\"subdir/index.html\\"]
      }
    </script>
  </head>
  <body></body>
</html>
"
`;

exports[`generate webpub from single markdown file 1`] = `
"/
└─ work/
   ├─ input/
   │  └─ foo.md
   └─ output/
      ├─ foo.html
      └─ publication.json"
`;

exports[`generate webpub from single markdown file 2`] = `
Object {
  "@context": Array [
    "https://schema.org",
    "https://www.w3.org/ns/pub-context",
  ],
  "conformsTo": "https://github.com/vivliostyle/vivliostyle-cli",
  "links": Array [],
  "name": "Hi",
  "readingOrder": Array [
    Object {
      "name": "Hi",
      "url": "foo.html",
    },
  ],
  "resources": Array [
    "publication.json",
  ],
  "type": "Book",
}
`;

exports[`generate webpub from vivliostyle.config.js 1`] = `
"/
└─ work/
   ├─ input/
   │  ├─ cover.html
   │  ├─ cover.png
   │  ├─ doc/
   │  │  ├─ one.html
   │  │  ├─ one.md
   │  │  ├─ two.html
   │  │  └─ two.md
   │  ├─ index.html
   │  ├─ output/
   │  │  ├─ cover.html
   │  │  ├─ cover.png
   │  │  ├─ doc/
   │  │  │  ├─ one.html
   │  │  │  └─ two.html
   │  │  ├─ index.html
   │  │  └─ publication.json
   │  ├─ publication.json
   │  └─ vivliostyle.config.json
   └─ output/
      ├─ cover.html
      ├─ cover.png
      ├─ doc/
      │  ├─ one.html
      │  └─ two.html
      ├─ index.html
      └─ publication.json"
`;

exports[`generate webpub from vivliostyle.config.js 2`] = `
Object {
  "@context": Array [
    "https://schema.org",
    "https://www.w3.org/ns/pub-context",
  ],
  "conformsTo": "https://github.com/vivliostyle/vivliostyle-cli",
  "links": Array [],
  "name": "output",
  "readingOrder": Array [
    Object {
      "rel": "cover",
      "type": "LinkedResource",
      "url": "cover.html",
    },
    Object {
      "name": "Table of Contents",
      "rel": "contents",
      "type": "LinkedResource",
      "url": "index.html",
    },
    Object {
      "url": "doc/one.html",
    },
    Object {
      "url": "doc/two.html",
    },
  ],
  "readingProgression": "rtl",
  "resources": Array [
    Object {
      "encodingFormat": "image/png",
      "name": "Cover image",
      "rel": "cover",
      "url": "cover.png",
    },
    "publication.json",
  ],
  "type": "Book",
}
`;

exports[`generate webpub from vivliostyle.config.js 3`] = `
"<html>
  <head>
    <meta charset=\\"utf-8\\" />
    <title>output</title>
    <link
      rel=\\"publication\\"
      type=\\"application/ld+json\\"
      href=\\"publication.json\\"
    />
  </head>
  <body>
    <h1>output</h1>
    <nav id=\\"toc\\" role=\\"doc-toc\\">
      <h2>Table of Contents</h2>
      <ol>
        <li><a href=\\"doc/one.html\\">one</a></li>
        <li><a href=\\"doc/two.html\\">two</a></li>
      </ol>
    </nav>
  </body>
</html>
"
`;

exports[`generate webpub with complex copyAsset settings 1`] = `
"/
└─ work/
   ├─ input/
   │  ├─ .vs/
   │  │  ├─ assetA.jxl
   │  │  ├─ doc.html
   │  │  ├─ node_modules/
   │  │  │  └─ pkgB/
   │  │  │     ├─ a.html
   │  │  │     └─ bar/
   │  │  │        └─ b.html
   │  │  └─ publication.json
   │  ├─ assetA.jxl
   │  ├─ assetB.svg
   │  ├─ cover.png
   │  ├─ doc.md
   │  ├─ node_modules/
   │  │  ├─ pkgA/
   │  │  │  └─ img.png
   │  │  └─ pkgB/
   │  │     ├─ a.html
   │  │     └─ bar/
   │  │        └─ b.html
   │  ├─ package.json
   │  └─ vivliostyle.config.json
   └─ output/
      ├─ assetA.jxl
      ├─ doc.html
      ├─ node_modules/
      │  └─ pkgB/
      │     ├─ a.html
      │     └─ bar/
      │        └─ b.html
      └─ publication.json"
`;
